# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: accessible/mac/MOXTextMarkerDelegate.mm
# Commit: 879f36821e41
# Full Hash: 879f36821e411d4a99399b9f5f4f48feeb7a1c8b
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2020-07-06 21:50:07
# Description:
#   Bug 1650740 - Don't allow text range retrieval when doc tree is not fully constructed. r=morgan
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D82410
# ==============================================================================

diff -r c69d97dae74d -r 879f36821e41 accessible/mac/MOXTextMarkerDelegate.mm
--- a/accessible/mac/MOXTextMarkerDelegate.mm	Mon Jul 06 17:21:37 2020 +0000
+++ b/accessible/mac/MOXTextMarkerDelegate.mm	Mon Jul 06 17:31:17 2020 +0000
@@ -60,6 +60,23 @@
 }
 
 - (NSString*)moxStringForTextMarkerRange:(id)textMarkerRange {
+  if (mGeckoDocAccessible.IsAccessible()) {
+    if (!mGeckoDocAccessible.AsAccessible()->AsDoc()->HasLoadState(
+            DocAccessible::eTreeConstructed)) {
+      // If the accessible tree is still being constructed the text tree
+      // is not in a traversable state yet.
+      return @"";
+    }
+  } else {
+    if (mGeckoDocAccessible.AsProxy()->State() & states::STALE) {
+      // In the proxy case we don't have access to load state,
+      // so we need to use the less granular generic STALE state
+      // this state also includes DOM unloaded, which isn't ideal.
+      // Since we really only care if the a11y tree is loaded.
+      return @"";
+    }
+  }
+
   mozilla::a11y::GeckoTextMarkerRange range(mGeckoDocAccessible, textMarkerRange);
   return range.Text();
 }
